Sources/Moya/TargetType.swift

```diff
@@ -38,32 +32,36 @@ public extension TargetType {
     }
 }
 
-/// Represents a type of upload task.
-public enum UploadType {
+/// Represents an HTTP task.
+public enum Task {
 
-    /// Upload a file.
-    case file(URL)
+    /// A request with no additional data.
+    case requestPlain
 
-    /// Upload "multipart/form-data"
-    case multipart([MultipartFormData])
-}
+    /// A requests body set with data.
+    case requestData(Data)
 
-/// Represents a type of download task.
-public enum DownloadType {
+    /// A requests body set with encoded parameters.
+    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
 
-    /// Download a file to a destination.
-    case request(DownloadDestination)
-}
+    /// A requests body set with data, combined with url parameters.
+    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
 
-/// Represents an HTTP task.
-public enum Task {
+    /// A requests body set with encoded parameters combined with url parameters.
+    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
```

Comment:
I was thinking of something more along the lines of making `urlRequest` a throwing method:
```swift 
func urlRequest() throws -> URLRequest
```
This way we can make use of the other encoding errors + add our own `MoyaError` cases.
I don't know how difficult this would be to implement without looking at the callers.

I'm not 100% against using a `fatalError` but I'll acknowledge I am a bit biased towards its use. It's certainly more explicit but it also feels like we're avoiding the problem here 😞